### PR TITLE
Solutions Chapter 2

### PR DESCRIPTION
**Pull request by L7R7**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -857,7 +874,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (a : as) = rewind as ++ [a]
```
Comment (chshersh) [inline src/Chapter2.hs]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Great solutions! Not much to comment :slightly_smiling_face: 
I left only one minor comment about the solution to the Advanced Task. Otherwise looks perfect :gem:

File: src/Chapter2.hs (hunk 1 above)
Comment (L7R7) [reply src/Chapter2.hs]: Ah, good point. I already suspected that the solution wasn't ideal in terms of performance, but I couldn't come up with a better solution in the first place. With your hint, I found a variant using the `go` pattern :smiley:

**Comment by L7R7** [comment]
Thank you for the review! I'm already working on Chapter 3 :muscle:

File: src/Chapter2.hs (hunk 1 above)
Comment (vrom911) [reply src/Chapter2.hs]: This is the absolutely right solution you have come up with for the `reverse` function and go pattern 👍🏼 Feel free to merge this PR 🚢

**Pull request merged**